### PR TITLE
Changing how we deal with Regularization

### PR DESCRIPTION
**Pull request by odunbar**
## Purpose 

Reduce the cost of regularizing with non-diagonal matrices dramatically. 
Updates many of the examples to work with the latest loss functions 

## Content

- Inverts the regularization matrix prior to solving the linear systems roughly speaking one solves `(A^T (reg)^{-1}A + I)x = reg^{-1}b` replacing  `(A^T A + reg)x = b`
- Updates `posdef_correct` to check if symmetrization is enough to ensure positive definiteness
- Updates examples to work with newer interfaces
- Updates `nd_to_md_regression_direct_withcov.jl` to match to test implementations of loss functions 
- Allow users to provide `inv(lambda)` so as not to internally invert the pxp matrix every time in, say, a hyperparameter learning scenario. set `regularization_inverted = true` flag in method constructor
- Update unit tests
## TODO

- possibly try out new weightings of the loss function in the examples - or leave this for future

## Example: 1d_to_3d_regression

`nd_to_md_regression_direct_withcov.jl` with updated loss and with new matrix solve, the training for the **non-diagonal** example with **non-diagonal regularization** takes 2.3secs and produces:

Compared with taking a **diagonal approximation of the regularization** is more prone to over-smoothing

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: examples/Learn_hyperparameters/nd_to_md_regression_direct_withcov.jl [hunk 1]
```diff
@@ -476,20 +455,18 @@ end
...
-        data = vcat(reshape(y[:, (n_train + 1):end], :, 1), 0.0, min_complexity) #flatten data
-    else
-        min_complexity = (n_features_opt / output_dim) * 2 * sum(log.(diag(cholesky(lambda).L)))
-        # TODO: SOLVE EVAL PROBLEM TO FIT WITH THIS DATA. the following is just a fudge to get reasonable scaling and not a global truth
-        min_complexity *= 1
+    #=
+    min_complexity =
+        isa(lambda, UniformScaling) ? n_features_opt * log(lambda.λ) :
+        n_features_opt / output_dim * 2 * sum(log.(diag(cholesky(lambda).L)))
+    min_complexity = sqrt(abs(min_complexity))
```
Comment (nickhnelsen) [inline examples/Learn_hyperparameters/nd_to_md_regression_direct_withcov.jl]: The `min_complexity` is now always equal to zero when using the new linear system with inverse data noise matrix. See bottom of page 14 on Overleaf.

File: examples/Learn_hyperparameters/nd_to_md_regression_direct_withcov.jl [hunk 2]
```diff
@@ -476,20 +455,18 @@ end
...
 
-        data = vcat(reshape(y[:, (n_train + 1):end], :, 1), 0.0, min_complexity) #flatten data
-    end
+
+    data = vcat(reshape(y[:, (n_train + 1):end], :, 1), 0.0, min_complexity) #flatten data
```
Comment (nickhnelsen) [inline examples/Learn_hyperparameters/nd_to_md_regression_direct_withcov.jl]: So, `min_complexity` in `data` should be replaced by `0.0`

File: examples/Learn_hyperparameters/nd_to_md_regression_direct_withcov.jl [hunk 3]
```diff
@@ -476,20 +455,18 @@ end
...
     println("min_complexity: ", min_complexity)
+    =#
+    data = zeros(size(Γ, 1))
```
Comment (nickhnelsen) [inline examples/Learn_hyperparameters/nd_to_md_regression_direct_withcov.jl:457]: Or maybe you are overwriting line 465 with line 468 here? In that case it seems correct as long as the forward map is defined by  `y[:, (n_train +1):end] - posterior_mean` in the first N times p rows of the map.

**Review by nickhnelsen — COMMENTED** [review]
@odunbar 

LGTM. I like the updated docs.

**Comment by odunbar** [comment]
@nickhnelsen Sorry for the confusion with the data. I'll delete the lines rather than comment them.

just as an addendum, this 1d->3d example isn't really good enough to discern between some of the loss functions. So I will see if it helps in the G-function example we had in CES package
